### DISCUSSION TITLE
tutorials/intro-typescript/13-using-models

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#126] [course feedback]
# tutorials/intro-typescript/13-using-models

https://www.apollographql.com/tutorials/intro-typescript/13-using-models

**Comment by roebuk** [comment]
I found this slightly confusing. You have the return type of the playlist as:

```
export type PlaylistModel = {
  id: string;
  name: string;
  description: string;
  tracks: {
    items: {
      track: TrackModel;
    }[];
  };
};
```


Which doesn't accurately represent the two possibilities that can be returned from the server. A more accurate representation, would be as follows:


```
export type PlaylistModel = {
  id: string;
  name: string;
  description: string;
  tracks: {
    items: PlaylistTrack | FeaturedPlaylistTrack
};

export type PlaylistTrack = {
  items: {
    track: TrackModel;
  }[];
};

export type FeaturedPlaylistTrack = {
  href: string;
  total: number;
};
```



The following would mean we have to slightly modify our `resolvers.ts` in order to satisfy TypeScript

```
function isPlaylistTrack(
  tracks: PlaylistTrack | FeaturedPlaylistTrack
): tracks is PlaylistTrack {
  return "items" in tracks;
}

export const resolvers: Resolvers = {
  Playlist: {
    tracks: async ({ tracks, id: playlistId }, _, { dataSources }) => {
      if (isPlaylistTrack(tracks)) {
        return tracks.items.map(({ track }) => track);
      }

      return dataSources.spotifyAPI.getTracks(playlistId);
    },
  },
};
```

**Reply by lizhennessy** [reply]
Hi there @roebuk ! This is a great catch. Previously, the `featured-playlists` endpoint _had_ returned full-fledged track objects; we just recently made the update to make follow-up requests necessary, and this evidently slipped through the cracks. We're working on a full course update that will better account for these types. Thanks for pointing it out!

**Reply by AlexRebula** [reply]
Hi there! Has this been addressed already? I would love to get an update. Thank you.

**Reply by AlexRebula** [reply]
Thank you for addressing this.
